src/tools/crawler/network.rs

```diff
@@ -0,0 +1,101 @@
+use std::{
+    cmp::Ordering,
+    collections::{HashMap, HashSet},
+    hash::{Hash, Hasher},
+    net::SocketAddr,
+    time::Duration,
+};
+
+use tokio::{sync::RwLock, time::Instant};
+use tracing::{error, info};
+
+#[derive(Default)]
+pub struct KnownNetwork {
+    nodes: RwLock<HashMap<SocketAddr, KnownNode>>,
+    connections: RwLock<HashSet<KnownConnection>>,
+}
+
+impl KnownNetwork {
+    pub(super) async fn insert_node(&self, addr: SocketAddr) {
+        let mut nodes = self.nodes.write().await;
+        if !nodes.contains_key(&addr) {
+            nodes.insert(addr, KnownNode::default());
+            info!("Known nodes: {}", nodes.len());
+        }
+    }
+
+    pub(super) async fn insert_connection(&self, addr: SocketAddr, peer: SocketAddr) {
+        let mut connections = self.connections.write().await;
+        connections.insert(KnownConnection::new(addr, peer));
+    }
+
+    pub(super) async fn set_connected(&self, addr: SocketAddr, handshake_time: Duration) {
+        let mut nodes = self.nodes.write().await;
+        if !nodes.contains_key(&addr) {
+            error!("set_connected(): addr not found ({})", addr);
+        } else {
+            nodes.get_mut(&addr).unwrap().handshake_time = Some(handshake_time);
+        }
+    }
+}
+
+/// A connection found in the network.
+#[derive(Debug, Eq, Copy, Clone)]
+pub struct KnownConnection {
+    /// One of the two sides of a connection.
+    pub a: SocketAddr,
+    /// The other side of a connection.
+    pub b: SocketAddr,
+    /// The timestamp of the last time the connection was seen.
+    pub last_seen: Instant,
+}
+
+impl Hash for KnownConnection {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        let (a, b) = (self.a, self.b);
+
+        // This ensures the hash is the same for (a, b) as it is for (b, a).
+        match a.cmp(&b) {
+            Ordering::Greater => {
+                b.hash(state);
+                a.hash(state);
+            }
+            _ => {
+                a.hash(state);
+                b.hash(state);
+            }
+        }
```

Comment:
maybe 
```suggestion
        if a.cmp(&b) == Ordering::Greater {
            b.hash(state);
            a.hash(state);
        } else {
            a.hash(state);
            b.hash(state);
        }
```

Reply:
That's the way it's done in zcash, so let's keep it consistent.